tests/migrate_signals/test_operation_signals.py

```diff
@@ -0,0 +1,155 @@
+
+    def _register(self, inject_operation):
+        self.addCleanup(
+            signals.post_operation.disconnect,
+            inject_operation,
+            sender=inject_operation.sender,
+        )
+
+    def test_apply_cyclical_infinite_recursion(self):
+        self._register(Receiver(Operation0, [Operation1()]))
+        self._register(Receiver(Operation1, [Operation1()]))
+
+        executor = MigrationExecutor(connection)
+
+        with self.assertRaisesMessage(
+            RecursionError, Migration.APPLY_RECURSION_ERROR_MESSAGE
+        ):
+            executor.migrate([("migrate_signals", "0001_initial")])
+
+    def test_apply_direct_infinite_recursion(self):
+        self._register(Receiver(Operation0, [Operation0()]))
+
+        executor = MigrationExecutor(connection)
+
+        with self.assertRaisesMessage(
+            RecursionError, Migration.APPLY_RECURSION_ERROR_MESSAGE
+        ):
+            executor.migrate([("migrate_signals", "0001_initial")])
+
+    def test_unapply_cyclical_infinite_recursion(self):
+        executor = MigrationExecutor(connection)
+        executor.migrate([("migrate_signals", "0001_initial")])
+        # Rebuild the graph to reflect the new DB state
+        executor.loader.build_graph()
+
+        self._register(Receiver(Operation0, [Operation1()]))
+        self._register(Receiver(Operation1, [Operation1()]))
+
+        with self.assertRaisesMessage(
+            RecursionError, Migration.APPLY_RECURSION_ERROR_MESSAGE
+        ):
+            executor.migrate([("migrate_signals", None)])
+
+    def test_unapply_direct_infinite_recursion(self):
+        executor = MigrationExecutor(connection)
+        executor.migrate([("migrate_signals", "0001_initial")])
+        # Rebuild the graph to reflect the new DB state
+        executor.loader.build_graph()
+
+        self._register(Receiver(Operation0, [Operation0()]))
+
+        with self.assertRaisesMessage(
+            RecursionError, Migration.APPLY_RECURSION_ERROR_MESSAGE
+        ):
+            executor.migrate([("migrate_signals", None)])
+
+    def test_order_forward(self):
+        self._register(Receiver(Operation0, [Operation1()]))
+        self._register(Receiver(Operation0, [Operation2()]))
+        self._register(Receiver(Operation0, [Operation1(), Operation2()]))
+        self._register(Receiver(Operation1, [Operation3()]))
+
+        applied_operations = []
+
+        def mock_run_python_state_forwards(self, app_label, state):
+            applied_operations.append(self)
+
+        executor = MigrationExecutor(connection)
+
+        with patch.object(
+            RunPython, "state_forwards", mock_run_python_state_forwards
+        ):
+            executor.migrate([("migrate_signals", "0001_initial")])
+            # Make sure we leave no trace of migrations behind for other tests.
+            self.addCleanup(
+                MigrationRecorder.Migration.objects.filter(app="migrate_signals").delete
+            )
```

Comment:
@charettes I had to add this cleanup on top of what you recommended. Is it the best way to do it? Or should I use something like the following?
```python
self.addCleanup(
    call_command,
    'migrate',
    'migrate_signals',
    'zero',
    database='default',
    interactive=False,
    verbosity=0,
)
```

Or use the executor?

```python
self.addCleanup(executor.loader.build_graph)
self.addCleanup(executor.migrate, [("migrate_signals", None)])
```

Reply:
I think `call_command` makes sense here.

Reply:
Wouldn't the 2 other methods be faster?

Reply:
Maybe, but they manipulate the internals a lot which shouldn't be necessary for such a test.